File: python/src/mshzip/coord_dict.py
'CoordDict — 적응형 XD 좌표 사전 (병렬 처리)'
from __future__ import annotations

import math
import os
from concurrent.futures import ThreadPoolExecutor

from . import hamming
from . import reed_solomon as rs

BITS_PER_AXIS = 1024
BYTES_PER_AXIS = 128       # 1024 / 8
HAMMING_BYTES = 130         # ceil(1035 / 8)
RS_GROUP_SIZE = 8
COORDDICT_EXTRA_HEADER_SIZE = 8
MIN_CHUNKS_FOR_PARALLEL = 4  # 이 이상일 때만 병렬화


class CoordDictPacker:
    """CoordDict 인코더 (C native + OpenMP 시 병렬 처리)."""

    def __init__(self, dimensions: int | None = None) -> None:
        self.dimensions = dimensions or os.cpu_count() or 4
        if self.dimensions < 1:
            raise ValueError(f'CoordDict dimensions must be >= 1, got {self.dimensions}')
        self.chunk_data_size = self.dimensions * BYTES_PER_AXIS
        self.rs_axes = math.ceil(self.dimensions / RS_GROUP_SIZE)
        self.total_axes_per_chunk = self.dimensions + self.rs_axes
        self.encoded_chunk_size = self.total_axes_per_chunk * HAMMING_BYTES
        self._executor: ThreadPoolExecutor | None = None

    def _get_executor(self) -> ThreadPoolExecutor:
        """Lazy 초기화: 첫 사용 시 생성."""
        if self._executor is None:
            self._executor = ThreadPoolExecutor(
                max_workers=min(os.cpu_count() or 4, 8)
            )
        return self._executor

    def encode(self, data: bytes | bytearray) -> tuple[bytes, int]:
        """입력 데이터를 CoordDict 인코딩. 반환: (encoded, seqCount)."""
        chunks: list[bytes] = []
        off = 0
        while off < len(data):
            chunk = data[off:off + self.chunk_data_size]
            if len(chunk) < self.chunk_data_size:
                chunk = chunk + b'\x00' * (self.chunk_data_size - len(chunk))
            chunks.append(chunk)
            off += self.chunk_data_size

        if not chunks:
            return b'', 0

        seq_count = len(chunks)

        # C native + 4청크 이상 → ThreadPoolExecutor 병렬 처리
        if len(chunks) >= MIN_CHUNKS_FOR_PARALLEL and hamming._USE_NATIVE:
            executor = self._get_executor()
            encoded_parts = list(executor.map(self.encode_chunk, chunks))
        else:
            encoded_parts = [self.encode_chunk(c) for c in chunks]

        return b''.join(encoded_parts), seq_count

    def close(self) -> None:
        """ThreadPoolExecutor 정리."""
        if self._executor:
            self._executor.shutdown(wait=False)
            self._executor = None

    def encode_chunk(self, chunk: bytes) -> bytes:
        """단일 청크 인코딩: D × 128B → (D + R) × 130B."""
        D = self.dimensions

        # D개 축으로 분배
        data_axes = [
            chunk[i * BYTES_PER_AXIS:(i + 1) * BYTES_PER_AXIS]
            for i in range(D)
        ]

        # Hamming 인코딩 (배치 API)
        hamming_axes = hamming.encode_batch(data_axes)

        # RS 패리티축
        parity_axes = rs.generate_parity(hamming_axes, RS_GROUP_SIZE)

        return b''.join(hamming_axes) + b''.join(parity_axes)


class CoordDictUnpacker:
    """CoordDict 디코더 (C native + OpenMP 시 병렬 처리)."""

    def __init__(self) -> None:
        self._executor: ThreadPoolExecutor | None = None

    def _get_executor(self) -> ThreadPoolExecutor:
        """Lazy 초기화."""
        if self._executor is None:
            self._executor = ThreadPoolExecutor(
                max_workers=min(os.cpu_count() or 4, 8)
            )
        return self._executor

    def decode(
        self,
        raw_payload: bytes,
        dimensions: int,
        rs_axes: int,
        seq_count: int,
        orig_bytes: int,
    ) -> bytes:
        """CoordDict 디코딩: 인코딩 페이로드 → 원본."""
        total_axes_per_chunk = dimensions + rs_axes
        encoded_chunk_size = total_axes_per_chunk * HAMMING_BYTES

        # 청크 슬라이스 분리
        encoded_chunks: list[bytes] = []
        offset = 0
        for _ in range(seq_count):
            encoded_chunks.append(raw_payload[offset:offset + encoded_chunk_size])
            offset += encoded_chunk_size

        # C native + 4청크 이상 → 병렬 디코딩
        if len(encoded_chunks) >= MIN_CHUNKS_FOR_PARALLEL and hamming._USE_NATIVE:
            executor = self._get_executor()
            decode_fn = lambda enc: self.decode_chunk(enc, dimensions, rs_axes)
            restored_chunks = list(executor.map(decode_fn, encoded_chunks))
        else:
            restored_chunks = [
                self.decode_chunk(enc, dimensions, rs_axes)
                for enc in encoded_chunks
            ]

        full = b''.join(restored_chunks)
        return full[:orig_bytes]

    def close(self) -> None:
        """ThreadPoolExecutor 정리."""
        if self._executor:
            self._executor.shutdown(wait=False)
            self._executor = None

    def decode_chunk(
        self,
        encoded_chunk: bytes,
        dimensions: int,
        rs_axes: int,
    ) -> bytes:
        """단일 청크 디코딩: (D + R) × 130B → D × 128B."""
        total_axes = dimensions + rs_axes

        # 축 분리
        all_axes = [
            encoded_chunk[i * HAMMING_BYTES:(i + 1) * HAMMING_BYTES]
            for i in range(total_axes)
        ]
        data_axes = all_axes[:dimensions]
        parity_axes = all_axes[dimensions:]

        # Hamming 디코딩 (배치 API)
        batch_results = hamming.decode_batch(data_axes)
        decoded = [r[0] for r in batch_results]
        damaged = [r[2] for r in batch_results]

        # RS 복구
        if any(damaged):
            recovered = rs.recover(data_axes, parity_axes, damaged, RS_GROUP_SIZE)
            for i in range(dimensions):
                if damaged[i]:
                    data, _, _ = hamming.decode(recovered[i])
                    decoded[i] = data

        return b''.join(decoded)

File: nodejs/lib/coord-dict.js
'use strict';

/**
 * CoordDict — 적응형 XD 좌표 사전
 *
 * 데이터를 D차원 좌표계로 변환, 축당 Hamming(1035,1024) ECC 적용.
 * D = CPU 코어 수에 자동 적응.
 */

const os = require('os');
const hamming = require('./hamming');
const rs = require('./reed-solomon');

const BITS_PER_AXIS = 1024;
const BYTES_PER_AXIS = 128;      // 1024 / 8
const HAMMING_BYTES = 130;        // ceil(1035 / 8)
const RS_GROUP_SIZE = 8;
const COORDDICT_EXTRA_HEADER_SIZE = 8;

class CoordDictPacker {
  /**
   * @param {Object} [opts]
   * @param {number} [opts.dimensions] - 차원 수 (기본: CPU 코어 수)
   */
  constructor(opts = {}) {
    this.dimensions = opts.dimensions || os.cpus().length;
    if (this.dimensions < 1) {
      throw new Error(`CoordDict dimensions must be >= 1, got ${this.dimensions}`);
    }
    this.chunkDataSize = this.dimensions * BYTES_PER_AXIS;
    this.rsAxes = Math.ceil(this.dimensions / RS_GROUP_SIZE);
    this.totalAxesPerChunk = this.dimensions + this.rsAxes;
    this.encodedChunkSize = this.totalAxesPerChunk * HAMMING_BYTES;
  }

  /**
   * 입력 데이터를 CoordDict 인코딩 (비압축)
   * @param {Buffer} input - 원본 데이터
   * @returns {{ encoded: Buffer, seqCount: number }}
   */
  encode(input) {
    // 청크 분할
    const chunks = [];
    for (let off = 0; off < input.length; off += this.chunkDataSize) {
      let chunk = input.slice(off, off + this.chunkDataSize);
      if (chunk.length < this.chunkDataSize) {
        const padded = Buffer.alloc(this.chunkDataSize, 0);
        chunk.copy(padded);
        chunk = padded;
      }
      chunks.push(chunk);
    }

    if (chunks.length === 0) {
      return { encoded: Buffer.alloc(0), seqCount: 0 };
    }

    const seqCount = chunks.length;

    // 각 청크 인코딩
    const encodedChunks = chunks.map(chunk => this.encodeChunk(chunk));

    return { encoded: Buffer.concat(encodedChunks), seqCount };
  }

  /**
   * 단일 청크 인코딩
   * @param {Buffer} chunk - D × 128바이트
   * @returns {Buffer} (D + R) × 130바이트
   */
  encodeChunk(chunk) {
    const D = this.dimensions;

    // D개 축으로 분배 (연속 슬라이스)
    const dataAxes = [];
    for (let i = 0; i < D; i++) {
      dataAxes.push(chunk.slice(i * BYTES_PER_AXIS, (i + 1) * BYTES_PER_AXIS));
    }

    // 각 축 Hamming 인코딩
    const hammingAxes = dataAxes.map(axis => hamming.encode(axis));

    // RS 패리티축 생성
    const parityAxes = rs.generateParity(hammingAxes, RS_GROUP_SIZE);

    // 모든 축 연결
    return Buffer.concat([...hammingAxes, ...parityAxes]);
  }
}

class CoordDictUnpacker {
  /**
   * CoordDict 디코딩
   * @param {Buffer} rawPayload - gzip 해제된 인코딩 데이터
   * @param {number} dimensions - D
   * @param {number} rsAxes - R
   * @param {number} seqCount - 총 청크 수
   * @param {number} origBytes - 원본 크기
   * @returns {Buffer} 원본 데이터
   */
  decode(rawPayload, dimensions, rsAxes, seqCount, origBytes) {
    const totalAxesPerChunk = dimensions + rsAxes;
    const encodedChunkSize = totalAxesPerChunk * HAMMING_BYTES;

    const restoredChunks = [];
    let offset = 0;

    for (let c = 0; c < seqCount; c++) {
      const encoded = rawPayload.slice(offset, offset + encodedChunkSize);
      offset += encodedChunkSize;

      restoredChunks.push(this.decodeChunk(encoded, dimensions, rsAxes));
    }

    const full = Buffer.concat(restoredChunks);
    return full.slice(0, origBytes);
  }

  /**
   * 단일 청크 디코딩
   * @param {Buffer} encodedChunk - (D + R) × 130바이트
   * @param {number} dimensions - D
   * @param {number} rsAxes - R
   * @returns {Buffer} D × 128바이트
   */
  decodeChunk(encodedChunk, dimensions, rsAxes) {
    const totalAxes = dimensions + rsAxes;

    // 축 분리
    const allAxes = [];
    for (let i = 0; i < totalAxes; i++) {
      allAxes.push(encodedChunk.slice(i * HAMMING_BYTES, (i + 1) * HAMMING_BYTES));
    }
    const dataAxes = allAxes.slice(0, dimensions);
    const parityAxes = allAxes.slice(dimensions);

    // Hamming 디코딩
    const decoded = [];
    const damaged = [];
    for (let i = 0; i < dimensions; i++) {
      const { data, corrected, uncorrectable } = hamming.decode(dataAxes[i]);
      decoded.push(data);
      damaged.push(uncorrectable);
    }

    // RS 복구 (손상 축이 있을 때)
    if (damaged.some(d => d)) {
      const recovered = rs.recover(dataAxes, parityAxes, damaged, RS_GROUP_SIZE);
      for (let i = 0; i < dimensions; i++) {
        if (damaged[i]) {
          const { data } = hamming.decode(recovered[i]);
          decoded[i] = data;
        }
      }
    }

    return Buffer.concat(decoded);
  }
}

module.exports = {
  CoordDictPacker,
  CoordDictUnpacker,
  BITS_PER_AXIS,
  BYTES_PER_AXIS,
  HAMMING_BYTES,
  RS_GROUP_SIZE,
  COORDDICT_EXTRA_HEADER_SIZE,
};
